### DISCUSSION TITLE
Application related to nonlinear optimization

### DISCUSSION THREAD

**Discussion by sandeep026** [symforce-org/symforce#377] [Q&A]
Dear developers,

I want to solve constrained nonlinear optimization problems using symforce for small to medium sized problems. Can I formulate objective , constraints and compute their derivatives for use in a solver like IPOPT? If there are some existing works, references will be greatly helpful.

Additional information
1. derivatives are usually sparse
2.  objectives and constraints can be evaluated in parallel

**Comment by ryan-brott-skydio** [comment]
> Can I formulate objective , constraints and compute their derivatives for use in a solver like IPOPT?
> 1. derivatives are usually sparse

Yes, SymForce can be used for these problems. We don't have any direct examples of this, but I'll sketch out how I would approach it.
* Break the constraints into dense fragments and [generate code](https://symforce.org/tutorials/codegen_tutorial.html) for each of them.
* Wrap these functions in `sym::Factor` objects and pass them to a `sym::Linearizer`. I would follow the [sparse linearizer test](https://github.com/symforce-org/symforce/blob/main/test/symforce_linearizer_test.cc).
You can do something similar for the objective (perhaps with the [dense linearizer](https://github.com/symforce-org/symforce/blob/main/test/symforce_dense_linearizer_test.cc) instead of the sparse one).

> 2. objectives and constraints can be evaluated in parallel

The SymForce linearizers do not exploit thread-level parallelism, though user code may strategically split factors between multiple linearizers, run linearization in parallel, and combine the linearizations together (one can imagine a few variants of this strategy).
